working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+leads to the desire to use _intersection types_ to express the property that
+some value is both a function of a specific type _and_ shareable:
+
+```dart
+class A implements Shareable {
+  void Function() f;  // compile time error
+  void Function() & Shareable f;  // ok
+}
+```
+
+Introducing intersection types into type system might be a huge undertaking. For
+the purposes of developing an MVP we can choose one of the two approaches:
+
+- Ignore functions entirely: consider functions un-shareable. It becomes a
+  compile time error to have a function type field inside a shareable class.
+- Allow function type fields inside a shareable class, but enforce shareability
+  in runtime on assignment to the field.
+
+### Shareable core types
+
+The following types will become shareable:
+
+- `num` (`int` and `double`), `String`, `bool`, `Null`, `BigInt`
+- `Enum` and consequently all user defined enums
+- `RegExp`, `DateTime`, `Uri`
+- `TypedData` - which makes all typed data types shareable.
+- `Pointer`
+- `Type` and `Symbol`
+- `StackTrace`
+
+#### Collections
+
+`dart:core` will provide shareable variants of all collection classes:
+
+```dart
+abstract interface class ShareableList<E extends Shareable?>
+    implements List<E>, Shareable {
+}
+
+abstract interface class ShareableSet<E extends Shareable?>
+    implements Set<E>, Shareable {
+}
+
+abstract interface class ShareableMap<K extends Shareable?,
+                                  V extends Shareable?>
+    implements Map<K, V>, Shareable {
+}
+```
+
+Default implementations of `List`, `Set` and `Map` will be changed to be
+shareable if their element type is shareable
+
+```dart
+<Shareable>[] is ShareableList<Shareable> // => true
+```
+
+We will also provide methods to convert collections to their shareable
+counterparts:
+
+```dart
+extension ToShareableList<E extends Shareable?> on List<E> {
+  /// Converts the list to [ShareableList] if it is not already shareable.
+  ShareableList<E> toShareable() =>
+    switch (this) {
+      final ShareableList<Shareable?> shareable => shareable,
+      _ => ShareableList<E>.from(this),
+    };
+}
+
+extension ToShareableSet<E extends Shareable?> on Set<E> {
+  ShareableSet<E> toShareable() => /* ... */
+}
+
+extension ToShareableMap<K extends Shareable?,
+                         V extends Shareable?> on Map<K, V> {
+  ShareableMap<K, V> toShareable() => /* ... */
+}
+```
+
+### `SendPort` semantics
+
+`SendPort.send` is extended with a named parameter to control whether shareable
+types are copied or shared with the receiver:
+
+```dart
+abstract interface class SendPort {
+    /// Sends an asynchronous [message] through this send port, to its
+    /// corresponding [ReceivePort].
+    ///
+    /// [...]
+    ///
+    /// If [copyShareables] is `false` but the sender and receiver
+    /// do not share the same code then an [IllegalArgument] exception
+    /// is thrown.
+    void send(Object? message, {bool copyShareables = true});
```

Comment:
I have initially considered `share` but I found it non-descriptive, consider for example:

```dart
class S implements Shareable {}

final arr = <Object>[S(), S()];
port.send(arr, share: true);
```

Should this actually throw an error? We are asking to _share_, but `arr` is not `Shareable` - it requires copying. On the other hand:

```dart
port.send(arr, copyShareables: false);
```

has clear semantics: `S` values will be shared, but `arr` itself will be copied (maintaining original `send` semantics). 

Alternatively we could do:

```
void share(Shareable message);
```

which also makes semantics clear.

Thoughts?

Reply:
> ```dart
> void share(Shareable message);
> ```

That's definitely clear.

`copyShareables: false` is ambiguous. The semantics could be (1) that if the `message` is not `Sharable` everything transitively is copied. However, it could also mean (2) that if the `message` is not `Sharable` but some of the reachable objects are, the reachable objects that are sharable will not be copied but shared.

I'm not convinced that users would want to have the share-if-possible semantics too often, it would make code hard to reason about. Users could still do that by `message is Sharable ? port.share(message) : port.send(message)` if they'd want to.

Reply:
Could also just flip the value and make it `shareShareable`.

I like the separate `share(Shareable)` because it's all or nothing. 
Copying until seeing a shareable, then sharing that, it's unpredictable for large graphs. It might be what you want for efficiency, but it's easy to over-share when all subexpressions are treated the same. If there is an option to not share a shareable (there doesn't have to be, they could always be forced shared), then it's because you may want not to, and then it's problematic that you can't control the granularity more precisely.

Reply:
Switched to `share()`.

Reply:
This is still `send`, did you push the commit that changes it to `share`?